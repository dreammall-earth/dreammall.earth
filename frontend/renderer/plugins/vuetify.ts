// eslint-disable-next-line import/no-unassigned-import
import '@mdi/font/css/materialdesignicons.css'
// eslint-disable-next-line import/no-unassigned-import
import 'vuetify/styles'
import { I18n, useI18n } from 'vue-i18n'
import { ThemeDefinition, createVuetify } from 'vuetify'
import { aliases as mdiAliases } from 'vuetify/iconsets/mdi'
import { VNumberInput } from 'vuetify/labs/VNumberInput'
import { createVueI18nAdapter } from 'vuetify/locale/adapters/vue-i18n'

import { aliases as customAliases } from '#assets/icons'
import dark from '#assets/scss/dark.module.scss'
import light from '#assets/scss/light.module.scss'

function makeThemeFromCssModule(theme: CSSModuleClasses, isDark: boolean): ThemeDefinition {
  return {
    dark: isDark,
    colors: {
      background: theme.backgroundColor,
      surface: theme.surfaceColor,
      primary: theme.primaryColor,
      secondary: theme.secondaryColor,
      success: theme.successColor,
      info: theme.infoColor,
      warning: theme.warningColor,
      error: theme.errorColor,
      font: theme.fontColor,
      'dropdown-background': theme.dropdownBackgroundColor,
      'cockpit-highlight': theme.cockpitHighlightColor,
      'dm-panel-submit-color': theme.dmPanelSubmitColor,
      'dm-panel-submit-background-color': theme.dmPanelSubmitBackgroundColor,
      'dm-panel-text-input-color': theme.dmPanelTextInputColor,
      'dm-panel-call-action-button-color': theme.dmPanelCallActionButtonColor,
      'dm-panel-call-action-button-background-color': theme.dmPanelCallActionButtonBackgroundColor,
      'dm-panel-call-action-button-indicator-color': theme.dmPanelCallActionButtonIndicatorColor,
      'dm-panel-call-action-button-indicator-background-color':
        theme.dmPanelCallActionButtonIndicatorBackgroundColor,
      'dm-panel-reminder-text-color': theme.dmPanelReminderTextColor,
      'dm-panel-reminder-link-color': theme.dmPanelReminderLinkColor,
      'dm-switch-inactive-color': theme.dmSwitchInactiveColor,
      'dm-switch-active-color': theme.dmSwitchActiveColor,
    },
    variables: {
      'border-color': theme.borderColor,
      'border-opacity': theme.borderOpacity,
      // Vuetify colors don't support opacity, so we use variables instead.
      // Beware that HEX values won't work.
      'icon-background': theme.iconBackground,
      'bottom-menu-background': theme.bottomMenuBackground,
      'sidebar-background': theme.sidebarBackground,
      'cockpit-input-background': theme.cockpitInputBackground,
      'cockpit-element-background': theme.cockpitElementBackground,
      'cockpit-chip-background': theme.cockpitChipBackground,
      'cockpit-chip-background-2': theme.cockpitChipBackground2,
      'panel-background': theme.panelBackground,
      'panel-element-background': theme.panelElementBackground,
      'dm-panel-text-input-background-color': theme.dmPanelTextInputBackgroundColor,
      'dm-panel-background-color': theme.dmPanelBackgroundColor,
      'dm-panel-border-color': theme.dmPanelBorderColor,
      'dm-panel-reminder-text-background-color': theme.dmPanelReminderTextBackgroundColor,
      'dm-panel-reminder-link-background-color': theme.dmPanelReminderLinkBackgroundColor,
    },
  }
}

const themes = {
  light: makeThemeFromCssModule(light, false),
  dark: makeThemeFromCssModule(dark, true),
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default (i18n: I18n<any, NonNullable<unknown>, NonNullable<unknown>, string, false>) =>
  createVuetify({
    components: {
      VNumberInput,
    },
    locale: {
      adapter: createVueI18nAdapter({ i18n, useI18n }),
    },
    ssr: true,
    icons: {
      aliases: { ...mdiAliases, ...customAliases },
    },
    theme: {
      defaultTheme: 'light',
      themes,
    },
  })
