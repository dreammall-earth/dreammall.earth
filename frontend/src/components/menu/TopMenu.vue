<template>
  <div class="top-menu mt-6 mt-sm-0">
    <v-app-bar flat class="app-bar" height="70">
      <v-row class="ma-1">
        <v-col class="d-none d-md-flex align-center">
          <a href="/" class="logo">
            <LogoImage size="small" />
          </a>
        </v-col>
        <v-col class="d-flex align-center justify-center">
          <TabControl />
        </v-col>
        <v-col class="d-none d-md-flex align-center">
          <v-row>
            <v-col class="d-flex align-center">
              <LightDarkSwitch class="d-none d-lg-flex" />
            </v-col>
            <v-col class="d-flex align-center justify-end">
              <Circle @click="toggleDrawer">
                <v-icon icon="$camera"></v-icon>
              </Circle>
              <UserInfo class="ml-2" />
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-app-bar>
  </div>
  <ListWithNavigationDrawer :drawer="drawer" :location="location" @update:drawer="updateDrawer" />
</template>

<script lang="ts" setup>
import { ref } from 'vue'

import ListWithNavigationDrawer from '#components/vuetify/Organisms/ListWithNavigationDrawer.vue'

import Circle from './CircleElement.vue'
import LightDarkSwitch from './LightDarkSwitch.vue'
import LogoImage from './LogoImage.vue'
import TabControl from './TabControl.vue'
import UserInfo from './UserInfo.vue'

const drawer = ref(false)
const toggleDrawer = () => {
  drawer.value = !drawer.value
}
const updateDrawer = (value: boolean) => {
  drawer.value = value
}
const location = ref<'bottom' | 'right' | 'left' | 'end' | 'top' | 'start'>('right')
</script>

<style scoped lang="scss">
.app-bar {
  position: sticky;
  top: 0;
  background: transparent !important;
}

.logo {
  width: 140px;
}

.top-menu {
  position: sticky;
  top: 0;
  z-index: 1000;
}
</style>
