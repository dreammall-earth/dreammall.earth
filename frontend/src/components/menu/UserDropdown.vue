<template>
  <ul class="dropdown border-sm pa-3 pb-2 my-2 bg-dropdown-background">
    <li>
      <button class="sign-out" @click="signOut">
        <v-icon icon="$logout"></v-icon>{{ $t('buttons.signout') }}
      </button>
    </li>
  </ul>
</template>

<script lang="ts" setup>
import { inject } from 'vue'

import GlobalErrorHandler from '#plugins/globalErrorHandler'
import AuthService from '#src/services/AuthService'

const authService = inject<AuthService>('authService')

async function signOut() {
  try {
    await authService?.signOut()
  } catch (error) {
    GlobalErrorHandler.error('auth error', error)
  }
}
</script>

<style scoped lang="scss">
.dropdown {
  font-size: 14px;
  list-style: none;
  border-radius: 20px;
}

.dropdown li {
  padding: 5px;
}

.dropdown button {
  display: flex;
  align-items: center;
  width: 100%;
}

.dropdown button i {
  margin-top: -2px;
  margin-right: 8px;
}
</style>
