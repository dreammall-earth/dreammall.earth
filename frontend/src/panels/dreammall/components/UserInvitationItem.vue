<template>
  <v-list-item class="user-invitation-item my-2 pa-2 pr-3 rounded-pill" elevation="0" rounded>
    <template #prepend>
      <v-avatar size="40">
        <v-img v-if="user.avatar" :src="user.avatar" :alt="user.name" cover></v-img>
        <svg v-else width="40" height="40" viewBox="0 0 40 40" xmlns="http://www.w3.org/2000/svg">
          <circle cx="20" cy="20" r="20" fill="#E0E0E0" />
          <circle cx="20" cy="15" r="7" fill="#BDBDBD" />
          <path
            d="M6 34.5C6 27.5964 11.5964 22 18.5 22H21.5C28.4036 22 34 27.5964 34 34.5V40H6V34.5Z"
            fill="#BDBDBD"
          />
        </svg>
      </v-avatar>
    </template>

    <v-list-item-title class="text-white font-weight-medium">
      {{ user.name }}
    </v-list-item-title>

    <template #append>
      <v-btn
        elevation="0"
        :color="user.invited ? 'success' : 'secondary'"
        class="invitation-button text-caption py-2 px-3 rounded-pill"
        size="small"
        @click="toggleInvitation"
      >
        {{
          user.invited
            ? $t('dream-mall-panel.setup.invitation-sent')
            : $t('dream-mall-panel.setup.invite')
        }}
      </v-btn>
    </template>
  </v-list-item>
</template>

<script lang="ts" setup>
import { defineProps, defineEmits } from 'vue'

import UserInvitation from '#src/panels/dreammall/interfaces/UserInvitation'

const props = defineProps<{
  user: UserInvitation
}>()

const emit = defineEmits<{
  (e: 'update:invited', value: boolean): void
}>()

const toggleInvitation = () => {
  emit('update:invited', !props.user.invited)
}
</script>

<style lang="scss" scoped>
.user-invitation-item {
  --background-user-item: #5d6670; // todo: save globally
  --background-invitation-button: #8b949b; // todo: save globally

  background-color: var(--background-user-item) !important;

  .invitation-button {
    &.bg-secondary {
      background-color: var(--background-invitation-button) !important;
    }
  }
}
</style>
