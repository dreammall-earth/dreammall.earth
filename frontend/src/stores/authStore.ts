import Cookies from 'js-cookie'
import { User } from 'oidc-client-ts'
import { acceptHMRUpdate, defineStore } from 'pinia'
import { computed, ref } from 'vue'

export const cookieStorage = {
  setItem(key: string, state: string) {
    Cookies.set('auth', state, { expires: 3 })
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getItem(key: string) {
    return Cookies.get('auth') || null
  },
} as Storage

export const useAuthStore = defineStore(
  'auth',
  () => {
    const user = ref<User | null>(null)

    const accessToken = computed(() => user.value?.access_token ?? '')
    const isLoggedIn = computed(() => !!user.value)

    const save = (u: User | null) => {
      user.value = u
    }

    const clear = () => {
      user.value = null
    }

    return {
      user,
      accessToken,
      isLoggedIn,
      save,
      clear,
    }
  },
  { persist: { storage: cookieStorage } },
)

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAuthStore, import.meta.hot))
}
