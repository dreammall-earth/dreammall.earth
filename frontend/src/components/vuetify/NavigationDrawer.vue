<template>
  <v-navigation-drawer v-model="drawer" :location="computedLocation" temporary width="400">
    <slot />
  </v-navigation-drawer>
</template>

<script lang="ts" setup>
import { computed, ref, watch } from 'vue'
import { useDisplay } from 'vuetify'

export type DrawerLocation = 'right' | 'bottom' | 'left' | 'end' | 'top' | 'start'

const props = withDefaults(
  defineProps<{
    modelValue: boolean
    location?: DrawerLocation
  }>(),
  {
    location: 'right',
  },
)

const emits = defineEmits(['update:modelValue'])
const display = useDisplay()
const drawer = ref(props.modelValue)

const computedLocation = computed(() => {
  return display.mobile.value ? 'bottom' : props.location
})

watch(
  () => props.modelValue,
  (newVal) => {
    drawer.value = newVal
  },
)

watch(
  () => drawer.value,
  (newVal) => {
    if (newVal !== props.modelValue) {
      emits('update:modelValue', newVal)
    }
  },
)
</script>
