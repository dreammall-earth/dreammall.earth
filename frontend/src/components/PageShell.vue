<template>
  <v-app>
    <slot />
  </v-app>
</template>

<script lang="ts" setup>
import TopMenu from '#components/menu/TopMenu.vue'
</script>
