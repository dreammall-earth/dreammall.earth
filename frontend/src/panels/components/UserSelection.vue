<template>
  <v-text-field
    v-model="userSearch"
    rounded
    clearable
    class="elevation-0 w-100"
    content-class="elevation-0"
    label="Suche"
    variant="solo-filled"
    append-inner-icon="mdi-magnify"
  />

  <div class="user-list-container w-100">
    <div v-if="error" class="text-center text-error">
      {{ $t('dream-mall-panel.setup.search-error') }}
    </div>

    <v-list
      v-if="displayedUsers.length"
      class="bg-transparent w-100 user-list"
      :class="{ loading: isLoading }"
    >
      <UserInvitationItem
        v-for="user in displayedUsers"
        :key="user.id"
        :user="user"
        @update:invited="updateInvitationStatus(user.id, $event)"
      />
    </v-list>

    <div v-if="!displayedUsers.length" class="text-center mt-4 text-grey-darken-2">
      {{ $t('dream-mall-panel.setup.no-result') }}
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, computed, watch } from 'vue'

import UserInvitationItem from '#src/panels/components/UserInvitationItem.vue'
import useSearchUsers from '#src/panels/composables/useSearchUsers'
import UserInvitation from '#src/panels/dreammall/interfaces/UserInvitation'

const users = defineModel<number[]>({ required: true })

const { searchUsers, searchResults, isLoading, error } = useSearchUsers()
const userSearch = ref('')

watch(userSearch, (newSearch) => {
  searchUsers(newSearch)
})

const displayedUsers = computed<UserInvitation[]>(() => {
  return searchResults.value.map((user) => ({
    id: user.id,
    name: user.name || user.username,
    avatar: null, // not supporting yet
    invited: users.value.includes(user.id), // todo: user already included in myTable?
  }))
})

const updateInvitationStatus = (userId: number, invited: boolean) => {
  let currentUsers: number[] = [...users.value]
  if (invited && !currentUsers.includes(userId)) {
    currentUsers.push(userId)
  } else if (!invited && currentUsers.includes(userId)) {
    currentUsers = currentUsers.filter((id) => id !== userId)
  }
  users.value = currentUsers
}
</script>

<style lang="scss" scoped>
.flat-text-field .v-field--variant-solo-filled {
  box-shadow: none !important;
}

.user-list-container {
  height: 200px;
  overflow-y: auto;

  .user-list {
    height: 100%;

    &.loading {
      opacity: 0.5;
      transition: opacity 0.5s ease;
    }
  }
}
</style>
