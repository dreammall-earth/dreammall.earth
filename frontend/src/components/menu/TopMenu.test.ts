import { mount, flushPromises } from '@vue/test-utils'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { h } from 'vue'
import { VApp } from 'vuetify/components'

import { useAuthStore } from '#stores/authStore'
import { authService } from '#tests/mock.authService'

import TopMenu from './TopMenu.vue'
import UserDropdown from './UserDropdown.vue'

describe('TopMenu', () => {
  const Wrapper = () => {
    return mount(VApp, {
      slots: {
        default: h(TopMenu),
      },
    })
  }

  let wrapper: ReturnType<typeof Wrapper>

  beforeEach(() => {
    wrapper = Wrapper()
  })

  it('renders', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('signout button', () => {
    const authStore = useAuthStore()

    const authServiceSpy = vi.spyOn(authService, 'signOut')
    const storeSpy = vi.spyOn(authStore, 'clear')

    beforeEach(() => {
      vi.clearAllMocks()
      authStore.save({
        access_token: 'access_token',
        profile: {
          aud: 'aud',
          sub: 'sub',
          exp: 1,
          iat: 1,
          iss: 'iss',
        },
        token_type: 'token_type',
        session_state: null,
        state: null,
        expires_at: new Date().valueOf() + 100,
        expires_in: 0,
        expired: false,
        scopes: ['email'],
        toStorageString: () => 'toStorageString',
      })
    })

    describe('without error', () => {
      beforeEach(async () => {
        await wrapper.find('button.user-info').trigger('click')
        await flushPromises()
        await wrapper.findComponent(UserDropdown).find('button.sign-out').trigger('click')
      })

      it('calls auth service sign out', () => {
        expect(authServiceSpy).toBeCalled()
      })

      it('clears the store', () => {
        expect(storeSpy).toBeCalled()
      })
    })

    describe('with error', () => {
      const consoleSpy = vi.spyOn(console, 'log')

      beforeEach(async () => {
        authServiceSpy.mockRejectedValue('Error!')
        await wrapper.find('button.user-info').trigger('click')
        await flushPromises()
        await wrapper.findComponent(UserDropdown).find('button.sign-out').trigger('click')
      })

      it('logs the error', () => {
        expect(consoleSpy).toBeCalledWith('auth error', 'Error!')
      })
    })
  })
})
