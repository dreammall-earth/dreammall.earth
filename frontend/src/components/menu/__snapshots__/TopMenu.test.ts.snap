// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopMenu > renders 1`] = `
<div
  class="v-application v-theme--light v-layout v-layout--full-height v-locale--is-ltr"
>
  <div
    class="v-application__wrap"
  >
    
    <div
      class="topmenu"
    >
      <header
        class="v-toolbar v-toolbar--flat v-toolbar--density-default v-theme--light v-locale--is-ltr v-app-bar py-4"
        style="top: 0px; z-index: 1004; transform: translateY(0%); position: fixed; transition: none !important; left: 0px; width: calc(100% - 0px - 0px);"
      >
        <!---->
        
        <div
          class="v-toolbar__content"
          style=""
        >
          <!---->
          <!---->
          
          <div
            class="v-row"
          >
            <div
              class="v-col d-flex align-center"
            >
              <a
                class="w-100 ml-8"
                href="/"
              >
                <div
                  class="v-responsive v-img v-img--booting w-100 logo-medium"
                  data-v-f6be2607=""
                >
                  <div
                    class="v-responsive__sizer"
                  />
                  
                  
                  <transition-stub
                    appear="true"
                    css="true"
                    name="fade-transition"
                    persisted="false"
                  >
                    <img
                      class="v-img__img v-img__img--contain"
                      src="/src/assets/dreammall-logo.svg"
                      style="display: none;"
                    />
                  </transition-stub>
                  <transition-stub
                    appear="false"
                    css="true"
                    name="fade-transition"
                    persisted="false"
                  >
                    <!---->
                  </transition-stub>
                  <!---->
                  <!---->
                  <!---->
                  
                  
                  <!---->
                </div>
              </a>
            </div>
            <div
              class="v-col d-flex justify-end"
            >
              <!--v-if-->
            </div>
          </div>
          
          <!---->
        </div>
        
        
        <transition-stub
          appear="false"
          css="true"
          name="expand-transition"
          persisted="false"
        >
          <!---->
        </transition-stub>
        
      </header>
    </div>
    
  </div>
</div>
`;
